src/storage/src/source/mysql/snapshot.rs

```diff
@@ -15,38 +15,70 @@
+//! time. Instead, MySQL allows us to perform a snapshot of a table and let us know at which point
+//! in time the snapshot was taken. Using this information we can take a snapshot at an arbitrary
+//! point in time and then rewind it to the desired `initial_gtid_set` by "rewinding" it. These two
+//! phases are described in the following section.
 //!
-//! To achieve this we must ensure that all workers snapshot the tables at the same GTID Set. We
-//! designate a snapshot leader worker that starts by acquiring a table lock on all tables to be
-//! snapshot and then reading the @@gtid_executed "GTID Set" at that point. Once the GTID Set is
-//! read, the leader sends a signal to all workers to start their transactions.
+//! ## Producing a snapshot at a known point in time.
 //!
-//! Each worker starts a new transaction with 'REPEATABLE READ' and 'CONSISTENT SNAPSHOT' semantics
-//! so that they can read a consistent snapshot of the table at the specific GTID Set the
-//! transaction was started from.
+//! Ideally we would like to start and transaction and ask MySQL to tell us the point in time this
+//! transaction is running at. As far as we know there isn't such API so we achieve this using
+//! table locks instead.
 //!
-//! Once all workers have started their transactions, they send a signal back to the leader to
-//! release the table locks. This ensures that all workers see a consistent view of the tables
-//! starting from the same GTID. The workers then read the snapshot data and publish it downstream.
+//! The full set of tables that are meant to be snapshotted are partitioned among the workers. Each
+//! worker initiates a connection to the server and acquires a table lock on all the tables that
+//! have been assigned to it. By doing so we establish a moment in time where we know no writes are
+//! happening to the tables we are interested in. After the locks are taken each worker reads the
+//! current upper frontier (`snapshot_upper`) using the `@@gtid_executed` system variable. This
+//! frontier establishes an upper bound on any possible write to the tables of interest until the
+//! lock is released.
+//!
+//! Each worker now starts a transaction via a new connection with 'REPEATABLE READ' and
+//! 'CONSISTENT SNAPSHOT' semantics. Due to linearizability we know that this transaction's view of
+//! the database must some time `t_snapshot` such that `snapshot_upper <= t_snapshot`. We don't
+//! actually know the exact value of `t_snapshot` and it might be strictly greater than
+//! `snapshot_upper`. However, because this transaction will only be used to read the locked tables
+//! and we know that `snapshot_upper` is an upper bound on all the writes that have happened to
+//! them we can safely pretend that the transaction's `t_snapshot` is *equal* to `snapshot_upper`.
+//! We have therefore succeeded in starting a transaction at a known point in time!
```

Comment:
I like how you've rewritten this -- it makes it much more clear!

Reply:
Thanks for the quick review!